### ISSUE TITLE
How to get blockNumber of event and Txhash of event

### ISSUE THREAD

**Issue by FahdW** [polkadot-js/api#3152]
```
// Import the API
const { ApiPromise } = require('@polkadot/api');

async function main () {
  // Create our API with a default connection to the local node
  const api = await ApiPromise.create();

  // Subscribe to system events via storage
  api.query.system.events((events) => {
    console.log(`\nReceived ${events.length} events:`);

    // Loop through the Vec<EventRecord>
    events.forEach((record) => {
      // Extract the phase, event and the event types
      const { event, phase } = record;
      const types = event.typeDef;

      // Show what we are busy with
      console.log(`\t${event.section}:${event.method}:: (phase=${phase.toString()})`);
      console.log(`\t\t${event.meta.documentation.toString()}`);

      // Loop through each of the parameters, displaying the type and data
      event.data.forEach((data, index) => {
        console.log(`\t\t\t${types[index].type}: ${data.toString()}`);
      });
    });
  });
}

main().catch((error) => {
  console.error(error);
  process.exit(-1);
});
```

Given this block from the recipe. How would we find the txhash for the events? and how would i identify which block this current set of events ended up in?

**Comment by jacogr** [comment]
Currently the subscriptions don't return the blockhash for the block where it applies. (There is non-breaking work ongoing to return this in addition)

Here is a cookbook entry to map extrinsics to their events - https://polkadot.js.org/docs/api/cookbook/blocks#how-do-i-map-extrinsics-to-their-events

**Comment by FahdW** [comment]
Thanks @jacogr, I experimented with that cookbook, the end goal would be able to pull the txHash for the extrinsic that was executed. Unfortunately running code like the one below crashes with some typing errors when a non system extrinsic is called but this might have to do with the way i set it up. That said if you cannot retrieve blocks/events/txhash while subscribing, is there anyway to retrieve it once you have the extrinsics
```
  await api.rpc.chain.subscribeNewHeads(async (header) => {
    const allRecords = await api.query.system.events.at(header.hash);
    const signedBlock = await api.rpc.chain.getBlock(header.hash);
    signedBlock.block.extrinsics.forEach(({ method: { method, section } }, index) => {
      const events = allRecords
        .filter(({ phase }) =>
          phase.isApplyExtrinsic &&
          phase.asApplyExtrinsic.eq(index)
        )
        .map(({ event }) => `${event.section}.${event.method}`);

      console.log(`${section}.${method}:: ${events.join(', ') || 'no events'}`);
    });
  })
```

**Comment by jacogr** [comment]
What crashes/fails?

The extrinsic itself doesn't contain the hash is appears at, only the header or block does.

Actually now that I see the above, there is also a derive that actually does all the logic above and returns extrinsics with their events. This one - https://github.com/polkadot-js/api/blob/master/packages/api-derive/src/chain/subscribeNewBlocks.ts eg. `api.derive.chain.subscribeNewBlocks(....)`

What is provides in the callback are these https://github.com/polkadot-js/api/blob/master/packages/api-derive/src/type/SignedBlockExtended.ts#L51

Basically each extrinsic is mapped like this - https://github.com/polkadot-js/api/blob/master/packages/api-derive/src/type/SignedBlockExtended.ts#L15-L20

dispatchInfo, dispatchError (as available), events, actual extrinsic.

It is no magic from what is above, just slightly expanded to try and consolidate everything.

**Comment by FahdW** [comment]
I will give this a go, and report back. As for the error the error seemed to have occurred during a specific extrinsic's that i believe are custom types. I am unsure if this is coming from the chain side of things and it requires us to fix things on the RPC side of it.

```
RPC-CORE: getBlock(hash?: BlockHash): SignedBlock:: createType(SignedBlock):: Struct: failed on block: {"header":"Header","extrinsics":"Vec<Extrinsic>"}:: Struct: failed on extrinsics: Vec<Extrinsic>:: createType(ExtrinsicV4):: createType(ExtrinsicSignatureV4):: Unable to create Enum via index 255, in Id, Index, Raw, Address32, Address20
```

```
REGISTRY: Error: Unable to create Enum via index 255, in Id, Index, Raw, Address32, Address20
```

**Comment by jacogr** [comment]
So in that case it seems to be the Address type. The API defaults to MultiAddress. The API FAQ has a section on how to adjust based on the chain. (There are 3 options that chains typically configure)

https://polkadot.js.org/docs/api/FAQ#i-cannot-send-transactions-sending-yields-decoding-failures

**Comment by FahdW** [comment]
Yep looks like that was the issue, after configuring the type it worked well. I am able to get the signer, the extrinsic executed and the tx hash, with the block number if I want. Just need to identify chain specific things but as far as this goes. I am able to do everything I need to. Thank you very much!
